src/types/src/parser.c

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright(c) 2006 to 2019 ADLINK Technology Limited and others
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Eclipse Distribution License
+ * v. 1.0 which is available at
+ * http://www.eclipse.org/org/documents/edl-v10.php.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
+ */
+#include <assert.h>
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+#include <stdbool.h>
+
+#define YYSTYPE DDS_TS_PARSER_STYPE
+#define YYLTYPE DDS_TS_PARSER_LTYPE
+
+#define YY_TYPEDEF_YY_SCANNER_T
+typedef void* yyscan_t;
+
+#define YY_NO_UNISTD_H 1 /* to surpress #include <unistd.h> */
+
+#include "parser.h"
+#include "idl.parser.h"
+#include "yy_decl.h"
+#include "idl.lexer.h"
+
+extern void dds_ts_stringify(dds_ts_node_t *root_node, char *buffer, size_t len);
+
+extern int
+dds_ts_parse_file(const char *file, void (*error_func)(int line, int column, const char *msg))
+{
+  int err = 0;
+  FILE *fh;
+
+  assert(file != NULL);
+
+  if ((fh = fopen(file, "rb")) == NULL) {
+    err = errno;
+    if (error_func != NULL)
```

Comment:
My beloved subject for playing the pedant! Thank you 😁

``NULL`` must be treated as a pointer-to-object because it is simply defined as a "null pointer constant", which in turn is defined as:

> An integer constant expression with the value 0, or such an expression cast to type void *, is called a null pointer constant.66) If a null pointer constant is converted to a pointer type, the resulting pointer, called a null pointer, is guaranteed to compare unequal to a pointer to any object or function.

and the possible definition ``(void *)0`` is a pointer-to-object.

The lovely little thing about C is that conversion between pointer-to-object and pointer-to-function is not guaranteed to work. (It is guaranteed in POSIX and on Windows, but certainly not on all platforms.) So technically, this _may_ result in an invalid conversion.

So technically, the ``NULL`` should be replaced by ``0``. The chances of this ever making any difference are smallish.